hera_cal/abscal.py

```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+        if constrain_model_to_data_ants:
+            # In order to avoid inflating to full interferometer dataset
+            # which will be unecessary for many earlier analyses
+            # prune model antennas to only include data antennas in hdd
+            # this will only work if the antennas in the baseline keys for each redundant group
+            # in the model are also present in the dataset so only use this if you are confident
+            # that this is the case.
+            all_data_ants = np.unique(np.hstack([hdd.ant_1_array, hdd.ant_2_array]))
+            all_model_ants = np.unique(np.hstack([hdm.ant_1_array, hdm.ant_2_array]))
+            assert np.all([ant in all_data_ants for ant in all_model_ants]), "All model antennas must be present in data if constrain_model_to_data_ants is True!"
+            hdm.select(antenna_nums=all_data_ants,
+                       keep_all_metadata=False)
+        hdm.inflate_by_redundancy()
+        # also make sure to only include baselines present in hdd.
+        hdm.select(bls=hdd.bls)
+
+    model, model_flags, model_nsamples = hdm.read()
+
+    data_ant_flags = synthesize_ant_flags(data_flags, threshold=ant_threshold)
+    model_ant_flags = synthesize_ant_flags(model_flags, threshold=ant_threshold)
+
+    lst_center = hdd.lsts[hdd.Ntimes // 2] * 12 / np.pi
+    field_str = 'LST={lst_center:%.2f} hrs'
+
+    if refant is None:
+        # dummy refant. Could be completely flagged for all we know.
+        refant_init = str(hdd.ant_1_array[0])
+    else:
+        refant_init = str(refant)
+    # initialize HERACal to store new cal solutions
+    hc = UVCal()
+    hc = hc.initialize_from_uvdata(uvdata=hdd, gain_convention='divide', cal_style='sky',
+                                   ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                   metadata_only=False, sky_field=field_str, cal_type='gain')
+    hc.write_calfits(output_filename, clobber=clobber)
+    hc = io.HERACal(output_filename)
+    hc.read()
+    hc.update(flags=data_ant_flags)
+    # generate cal object from model to hold model flags.
+    hcm = UVCal()
+    hcm = hcm.initialize_from_uvdata(uvdata=hdm, gain_convention='divide', cal_style='sky',
+                                     ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                     metadata_only=False, sky_field=field_str, cal_type='gain')
+    hcm.write_calfits(output_filename + '.tmp')
+    hcm = io.HERACal(output_filename + '.tmp')
+    hcm.read()
+    hcm.update(flags=model_ant_flags)
+    # init all gains to unity.
+    hcm.gain_array[:] = 1. + 0.j
+    hc.gain_array[:] = 1. + 0.j
+    # set calibration flags to be or of data and model flags
+    hc.flag_array = hc.flag_array | hcm.flag_array
+
+    abscal_gains, abscal_flags, _, _ = hc.build_calcontainers()
+
+    # load in weights if supplied.
+    if auto_file is not None:
+        hda = io.HERAData(auto_file)
```

Comment:
so `to_HERAData` here

Reply:
changed.

Reply:
I've found that using strings in `to_HERAData` causes some very strange issues.

```
hera_cal/tests/test_abscal.py Fatal Python error: Illegal instruction

Current thread 0x000000010de52600 (most recent call first):
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/aipy_extracts.py", line 295 in __init__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/miriad.py", line 837 in read_miriad
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 10623 in read_miriad
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 12164 in read
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 11826 in read
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/io.py", line 535 in read
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/abscal.py", line 3748 in run_model_based_calibration
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/tests/test_abscal.py", line 967 in test_run_model_based_calibration
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/python.py", line 1641 in runtest
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 311 in from_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 323 in _main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/config/__init__.py", line 162 in main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/Users/aewallwi/miniconda3/envs/hera/bin/pytest", line 8 in <module>
Illegal instruction: 4
```